### PR TITLE
Add TMA tiling utils into creed

### LINKED ISSUE #12: Add tiling scripts to the repo
Add the tiling notebooks currently in `ark` along with `tiling_utils.py` over.

### PR DESCRIPTION
**Pull request by alex-l-kong**
Closes #12. Now that the tiling processes have been finalized, we add them to `creed-helper` to ensure they can be accessed on the MibiScope computer.

We'll start with the TMA tiling utils first, then add the tiled regions/non-TMA in afterward.

### CODE REVIEW

**Comment by alex-l-kong** [comment]
Copied over the `tiling_utils.py` and `example_tma_fov_generate.ipynb` over. I'll be working on setting up `Docker` and `Travis` separately; `base_dir` in the notebook is just a dummy value for the time being.

**Comment by ngreenwald** [comment]
travis/pytest/coveralls yes, docker no, since this will be on windows machines

File: creed/tiling_utils_test.py [hunk 1]
```diff
@@ -0,0 +1,755 @@
...
+        )
+        assert all(
+            sample_region_params[i]['y_fov_size'] == 1 for i in
+            range(len(sample_region_params))
+        )
+
+        # assert randomize is set to Y for both fovs
+        assert all(
+            sample_region_params[i]['region_rand'] == 'Y' for i in
+            range(len(sample_region_params))
```
Comment (ngreenwald) [inline creed/tiling_utils_test.py:275]: same thing here

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Looking good. A few more naming suggestions

File: creed/tiling_utils.py [hunk 2]
```diff
@@ -0,0 +1,1123 @@
...
+    return region_params_list
+
+
+def tiled_region_read_input(fov_list_info, region_params):
+    """Reads input for tiled regions from user and `fov_list_info`.
+
+    Updates all the tiling params inplace. Units used are microns.
+
+    Args:
+        fov_list_info (dict):
```
[the remark this answers is not in the document]
Comment (alex-l-kong) [reply creed/tiling_utils.py:126]: I think `region_corners` sounds good.

File: creed/tiling_utils.py [hunk 3]
```diff
@@ -0,0 +1,1123 @@
...
+            lambda ny: ny >= 1,
+            dtype=int
+        )
+
+        region_params['fov_num_x'].append(num_x)
+        region_params['fov_num_y'].append(num_y)
+
+        # allow the user to specify the step size along each dimension
+        size_x = read_tiling_param(
+            "Enter the x step size for region %s (in microns): " % fov['name'],
```
[the remark this answers is not in the document]
Comment (alex-l-kong) [reply creed/tiling_utils.py:157]: Are you talking about the `frameSizePixels` key? If that's the case, a couple follow-ups:

1. Does the commercial instrument take `frameSizeMicrons` as well or does it need to be `frameSizePixels`?
2. More to the above point, if `frameSizePixels` is the only supported spec, using the fiducial constants in reverse to convert pixels to microns should work.
3. The above points are for tiled regions (non-TMA); how will this affect TMA tiling, if at all? Can we ignore `frameSizePixels` and assume the user will just want to specify a specific number of FOVs along each coordinate axis, or should we determine programmatically how many FOVs along the x- and y-axes we should use based on `frameSizePixels` (or `frameSizeMicrons` if that can be specified)?

File: creed/tiling_utils.py [hunk 4]
```diff
@@ -0,0 +1,1123 @@
...
+        lambda mii: mii in ['Y', 'N', 'y', 'n'],
+        dtype=str
+    )
+
+    # convert to uppercase to standardize
+    moly_interval_insert = moly_interval_insert.upper()
+
+    # if moly insert is set, we need to specify an additional moly_interval param
+    # NOTE: the interval applies regardless of if the fovs overlap regions or not
+    if moly_interval_insert == 'Y':
```
[the remark this answers is not in the document]
Comment (alex-l-kong) [reply creed/tiling_utils.py:263]: How does something like this look in comparison:

```
`moly_interval = read_tiling_param(
    "Enter the FOV interval size to insert Moly points. If you do not want to insert Moly points at a FOV interval, enter 0.",
    "Error: moly interval must be 0 or a positive integer",
    lambda mi: mi >= 0,
    dtype=int
)

if moly_interval > 0:
    tiling_params['moly_interval'] = moly_interval
```

File: creed/tiling_utils.py [hunk 5]
```diff
@@ -0,0 +1,1123 @@
...
+        moly_interval = read_tiling_param(
+            "Enter the FOV interval size to insert moly points: ",
+            "Error: moly interval must be a positive integer",
+            lambda mi: mi >= 1,
+            dtype=int
+        )
+
+        tiling_params['moly_interval'] = moly_interval
+
+    return tiling_params, moly_point
```
[the remark this answers is not in the document]
Comment (alex-l-kong) [reply creed/tiling_utils.py:273]: We had agreed to do it this way when the tiling notebook was originally spun up, but I can change it back to loading it in the function.

**Comment by ngreenwald** [comment]
I changed my mind, we should add back the saving of the automatic TMA coords to make it easier to reproduce the output at a later date

File: creed/tiling_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ngreenwald) [reply creed/tiling_utils.py:157]: Lets discuss at standup

File: creed/tiling_utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ngreenwald) [reply creed/tiling_utils.py:251]: looks good

File: creed/tiling_utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ngreenwald) [reply creed/tiling_utils.py]: Which one do you think makes more sense?

[the remark this answers is not in the document]
Comment (alex-l-kong) [reply creed/tiling_utils.py]: To me, it makes more sense to load it in `generate_tiled_region_fov_list`. It's a little awkward as it is right now and this ensures we only load in the Moly point when we actually need it. It also matches up with what we do in `remap_and_reorder_fovs`.

**Comment by alex-l-kong** [comment]
> I changed my mind, we should add back the saving of the automatic TMA coords to make it easier to reproduce the output at a later date

To clarify, you mean creating a full-fledged JSON run file containing each automatic TMA coord like we did earlier? Or would simply saving the mapping from automatic FOV name be coordinate enough?

**Comment by ngreenwald** [comment]
Saving the current pared down automatic JSON that just contains names and centroid coordinates. 

Basically, we want someone in 3 months to be able to recreate the same interface. So they'll need to load their manual names, the automatic names we created, and the final mapping they decided on, so they can inspect the results

File: creed/tiling_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (alex-l-kong) [reply creed/tiling_utils.py:157]: Per our standup today, we'll open a separate issue to handle this after the TMA process is merged in. We'll be using `fovSizeMicrons` to space the tiled regions (non-TMA) version.

File: creed/tiling_utils.py [hunk 6]
```diff
@@ -0,0 +1,1115 @@
...
+
+    # need to copy fov metadata over, needed for generate_fov_list
+    tiling_params['fovs'] = copy.deepcopy(tiled_region_corners['fovs'])
+
+    # store the read in parameters in the region_params key
+    tiling_params['region_params'] = generate_region_info(region_params)
+
+    # whether to insert moly points between regions
+    moly_region_insert = read_tiling_param(
+        "Insert moly points between regions? Y/N: ",
```
Comment (ngreenwald) [inline creed/tiling_utils.py:231]:
```suggestion
        "Insert a moly point between each tiled region? Y/N: ",
```

File: creed/tiling_utils.py [hunk 7]
```diff
@@ -0,0 +1,1115 @@
...
+        dtype=str
+    )
+
+    # convert to uppercase to standardize
+    moly_region_insert = moly_region_insert.upper()
+    tiling_params['moly_region'] = moly_region_insert
+
+    # whether to insert moly points between fovs
+    moly_interval = read_tiling_param(
+        "Enter the FOV interval size to insert Moly points. If you do not want to insert "
```
Comment (ngreenwald) [inline creed/tiling_utils.py:243]: Change to: "Insert moly points between FOVs? If yes, enter the number of FOVs between each Moly point. If no, enter 0. "

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Looks good, minor comments only

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Looks good, re-request once the rhombus branch is merged in

**Comment by alex-l-kong** [comment]
After we merge this into `master`, I'll update the `tma_validate` PR to merge directly into `main`.

File: creed/tiling_utils.py [hunk 8]
```diff
@@ -0,0 +1,1180 @@
...
+import matplotlib.pyplot as plt
+import numpy as np
+from operator import itemgetter
+import os
+from skimage.draw import ellipse
+from sklearn.utils import shuffle
+
+from dataclasses import dataclass
+
+from creed import settings
```
Comment (ngreenwald) [inline creed/tiling_utils.py:15]: merge main into this branch update creed references to toffy

File: creed/tiling_utils_test.py [hunk 9]
```diff
@@ -0,0 +1,745 @@
...
+        'fov_num_y': [6, 12],
+        'x_fov_size': [5, 10],
+        'y_fov_size': [10, 20],
+        'region_rand': ['N', 'Y']
+    }
+
+    # generate the region params
+    sample_region_params = tiling_utils.generate_region_info(sample_region_inputs)
+
+    # assert region_start_x set correctly
```
Comment (ngreenwald) [inline creed/tiling_utils_test.py:154]: These assert statements can be condensed into a for loop over the tuple of parameter (i.e. region_start_x), base_value (i.e. 100)

File: creed/tiling_utils_test.py [hunk 10]
```diff
@@ -0,0 +1,745 @@
...
+        sample_region_params[i]['y_fov_size'] == 10 * (i + 1)
+        for i in range(len(sample_region_params))
+    )
+
+    # assert region_rand set correctly
+    assert sample_region_params[0]['region_rand'] == 'N'
+    assert sample_region_params[1]['region_rand'] == 'Y'
+
+
+@pytest.mark.parametrize('moly_interval_val', test_cases._PARAM_SET_MOLY_INTERVAL_VALUE_CASES)
```
Comment (ngreenwald) [inline creed/tiling_utils_test.py:195]: add a `parameterize = pytest.mark.paramtrize` at the top to shorten calls

File: creed/tiling_utils_test.py [hunk 11]
```diff
@@ -0,0 +1,745 @@
...
+        dtype=str
+    )
+
+    # assert sample_tiling_param was set to 'Y'
+    assert sample_tiling_param == 'Y'
+
+
+def test_read_tiled_region_inputs(monkeypatch):
+    # define a sample fovs list
+    sample_fovs_list = test_utils.generate_sample_fovs_list(
```
Comment (ngreenwald) [inline creed/tiling_utils_test.py:106]: This function should be moved from ark to this repo.

File: creed/tiling_utils_test.py [hunk 12]
```diff
@@ -0,0 +1,745 @@
...
+
+    # override the default functionality of the input function
+    monkeypatch.setattr('builtins.input', lambda _: next(user_inputs))
+
+    # use the dummy user data to read values into the params lists
+    tiling_utils.read_tiled_region_inputs(
+        sample_fovs_list, sample_region_params
+    )
+
+    assert (sample_region_params[i]['region_start_x'] == 100 * i
```
Comment (ngreenwald) [inline creed/tiling_utils_test.py:124]: This should be a for loop. In addition, the values that are being evaluated should be passed from the function that creates the data, rather than being hardcoded here

File: creed/tiling_utils_test.py [hunk 13]
```diff
@@ -0,0 +1,745 @@
...
+    )
+
+    # assert fov_num_y set correctly
+    assert all(
+        sample_region_params[i]['fov_num_y'] == 6 * (i + 1)
+        for i in range(len(sample_region_params))
+    )
+
+    # assert x_fov_size set correctly
+    assert all(
```
Comment (ngreenwald) [inline creed/tiling_utils_test.py:179]: Same thing here

File: creed/tiling_utils_test.py [hunk 14]
```diff
@@ -0,0 +1,745 @@
...
+        # assert the fovs in the tiling params are the same as in the original fovs list
+        assert sample_tiling_params['fovs'] == sample_fovs_list['fovs']
+
+        # assert region start x and region start y values are correct
+        sample_region_params = sample_tiling_params['region_params']
+        fov_0 = sample_fovs_list['fovs'][0]
+        fov_1 = sample_fovs_list['fovs'][1]
+
+        # assert region start, fov_num, and fov_size set correctly
+        assert (sample_region_params[i]['region_start_x'] == 100 * i
```
Comment (ngreenwald) [inline creed/tiling_utils_test.py:231]: Same thing here.

File: creed/tiling_utils_test.py (hunk 10 above)
Comment (ngreenwald) [reply creed/tiling_utils_test.py:195]: Given that this global is only used for this function, it makes more sense to just use [0, 1] here instead

File: creed/tiling_utils_test_cases.py [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+    def case_x_slant(self):
+        coords = [
+            XYCoord(1500, 2000),
+            XYCoord(2000, 2000),
+            XYCoord(1600, 1000),
+            XYCoord(2200, 1000)
+        ]
+
+        # total x-offset is 150 (average of the left-offset (100) and right-offset (200))
+        # 3 fovs along the y-axis, so divide by 2 to get individual fov offset (75)
```
Comment (ngreenwald) [inline creed/tiling_utils_test_cases.py:116]: This comment is useful, but it would be better to define the x_increment using the relationship you describe, so that if coords changes this value changes automatically too

File: creed/tiling_utils_test.py [hunk 16]
```diff
@@ -0,0 +1,745 @@
...
+
+@parametrize_with_cases('top_left, top_right, bottom_left, bottom_right',
+                        cases=test_cases.ValidateRhombusCoordsTests, glob='*_failure')
+def test_validate_tma_corners_failure(top_left, top_right, bottom_left, bottom_right):
+    # error checks 1-4: test invalid top_left, top_right, bottom_left, bottom_right respectively
+    with pytest.raises(ValueError):
+        tiling_utils.validate_tma_corners(top_left, top_right, bottom_left, bottom_right)
+
+
+@parametrize_with_cases('top_left, top_right, bottom_left, bottom_right',
```
Comment (ngreenwald) [inline creed/tiling_utils_test.py:413]: You can mark the cases that are expected to fail with `xfail`; see the rosetta PR. This will allow you to combine these functions together

File: creed/tiling_utils_test.py [hunk 17]
```diff
@@ -0,0 +1,745 @@
...
+    )
+
+    with tempfile.TemporaryDirectory() as td:
+        sample_tma_corners_path = os.path.join(td, 'sample_tma_corners.json')
+
+        with open(sample_tma_corners_path, 'w') as sfl:
+            json.dump(sample_fovs_list, sfl)
+
+        # test 1: tma_corners_file does not exist
+        # test 2: invalid num_x
```
Comment (ngreenwald) [inline creed/tiling_utils_test.py:436]: This should be in the cases documentation, rather than here

File: creed/tiling_utils_test_cases.py [hunk 18]
```diff
@@ -0,0 +1,214 @@
+import numpy as np
+
+from creed.tiling_utils import XYCoord
+
+# for tiled region param dict creation
+_PARAM_SET_MOLY_INTERVAL_VALUE_CASES = [0, 1]
+
+
+# for tiled region FOV generation
+class TiledRegionRandomizeCases:
```
Comment (ngreenwald) [inline creed/tiling_utils_test_cases.py:10]: Putting something in the cases file is useful to hide complicated functions. For simple lists of values, having them in the decorator is better, because than its easy to see what's being tested

File: creed/tiling_utils_test_cases.py [hunk 19]
```diff
@@ -0,0 +1,214 @@
...
+        return ['N', 'N']
+
+    def case_fov2_random(self):
+        return ['N', 'Y']
+
+    def case_both_random(self):
+        return ['Y', 'Y']
+
+
+class TiledRegionMolySettingCases:
```
Comment (ngreenwald) [inline creed/tiling_utils_test_cases.py:21]: this is a good example one that does belong in here

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Make sure you merge main into this branch to get the toffy changes. 

Most of my comments are about the testing. Feel free to address those in the validation PR if you'd prefer to merge this in, or do it in here if it'll be easier to keep track of

File: creed/tiling_utils_test.py (hunk 9 above)
Comment (alex-l-kong) [reply creed/tiling_utils_test.py]: Were you thinking of something more or less like:

```
# define the starting values to increment from for each param
base_vals = {
    'region_start_x': 100,
    'region_start_y': 200,
    'fov_num_x': 3,
    'fov_num_y': 6,
    'x_fov_size': 5,
    'y_fov_size': 10
}

# the keys are going to be the same for all regions of sample_region_params
for i in list(zip(sample_region_params[0].items()))
    for param, val in list(sample_region_params[i].items()):
        assert val == base_vals[param] * (i + 1)
```

Comment (ngreenwald) [reply creed/tiling_utils_test.py]: Yup, and you could use base_vals to generate sample_region_inputs as well

**Comment by alex-l-kong** [comment]
Ok, refactoring's done. Had a couple TODO questions but everything else has been addressed.

I pulled out some tests from `parametrize_with_cases` to `parametrize` with a specific `marks=[xfail]` argument where it was easier to test that way.

**Review by ngreenwald — APPROVED** [review]
Sweet, looks good
